operators/multiclusterobservability/controllers/placementrule/placementrule_controller.go

```diff
@@ -514,44 +516,74 @@ func updateManagedClusterList(obj client.Object) {
 	}
 }
 
+// Do not reconcile objects if this instance of mch has the
+// `disableAddonAutomaticInstallationAnnotationKey` annotation
+func isAutomaticAddonInstallationDisabled(obj client.Object) bool {
+	annotations := obj.GetAnnotations()
+	if val, ok := annotations[disableAddonAutomaticInstallationAnnotationKey]; ok && strings.EqualFold(val, "true") {
+		log.Info("Cluster has disable addon automatic installation annotation. Skip addon deploy")
+		return true
+	}
+	return false
+}
+
 // SetupWithManager sets up the controller with the Manager.
 func (r *PlacementRuleReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	c := mgr.GetClient()
 	ingressCtlCrdExists := r.CRDMap[config.IngressControllerCRD]
 	clusterPred := predicate.Funcs{
 		CreateFunc: func(e event.CreateEvent) bool {
 			log.Info("CreateFunc", "managedCluster", e.Object.GetName())
+
 			updateManagedClusterList(e.Object)
 			updateManagedClusterImageRegistry(e.Object)
+
+			if isAutomaticAddonInstallationDisabled(e.Object) {
+				return false
+			}
+
 			return true
 		},
 		UpdateFunc: func(e event.UpdateEvent) bool {
 			log.Info("UpdateFunc", "managedCluster", e.ObjectNew.GetName())
-			if e.ObjectNew.GetResourceVersion() != e.ObjectOld.GetResourceVersion() {
-				if e.ObjectNew.GetDeletionTimestamp() != nil {
-					log.Info("managedcluster is in terminating state", "managedCluster", e.ObjectNew.GetName())
-					managedClusterListMutex.Lock()
-					delete(managedClusterList, e.ObjectNew.GetName())
-					managedClusterListMutex.Unlock()
-					managedClusterImageRegistryMutex.Lock()
-					delete(managedClusterImageRegistry, e.ObjectNew.GetName())
-					managedClusterImageRegistryMutex.Unlock()
-				} else {
-					updateManagedClusterList(e.ObjectNew)
-					updateManagedClusterImageRegistry(e.ObjectNew)
-				}
-				return true
+
+			if e.ObjectNew.GetResourceVersion() == e.ObjectOld.GetResourceVersion() {
```

Comment:
I reorganized here in favor of early returns. Could the disable addon function also be an early return, or is it preferred to be at the bottom to allow other information to propegate?